### ISSUE TITLE
Registration - Put the scripts/data to run Rohlfing experiment on some standard place

### ISSUE THREAD

**Issue by omarocegueda** [dipy/dipy#441]
Rohlfing's experiment consists on registering 306 pairs of manually annotated MRI (IBSR) and measure the quality of the registration in terms of the average Jaccard index of each anatomical region. Right now, I am running this test on a cluster in my institution when we need to. We need to put the scripts and the data to easily run the experiment on some standard place. Each registration takes about 15 minutes.

**Comment by matthew-brett** [comment]
I think we should write to the INCF to see if they can provide a service like this.  I think it would be useful for anyone evaluating a registration method.

@jbpoline, @chrisfilo - what do you think?  How should we go about this?

**Comment by arokem** [comment]
Sorry for only picking up on this now. We can host data (and also scripts), with semantic versioning, on the Stanford SDR, if that seems like a good solution. People would still have to download the data/scripts to run them on some local resource, though.

**Comment by matthew-brett** [comment]
Right - I was thinking that we would need some system which would run something like - say 50 jobs at a time? Omar - is your estimate then 15 minutes * 306 for total CPU minutes?  What CPUs are you running on?  What kind of memory do they need?

**Comment by chrisfilo** [comment]
Let me try to understand the problem better: 
a) you need to host the manually annotated data, example scripts to show how to validate an algorithm using the data?
b) you want to build a Registration Evaluation as a Service platform which will allow people to submit their coregistration algorithms (packaged in some way) and get an evaluation of their performance.

I can help with a). b) will be a substantial effort in terms of software development if built ground up. However, it could become a part of a bigger Science as a Service platform.

**Comment by matthew-brett** [comment]
Omar knows this much better than me, but I believe our immediate need is to be able to run scripts, say hosted on github, that will execute a multi-CPU job on the a given standard set of data.

So yes, we would need:

1) somewhere to put the original data where the cluster can see it (and it would be better if this was also public;
2) CPUs on which to run the scripts, perhaps every week or so;
3) short term hosting of the output registrations and registration measures, which we could then download somewhere.

Being able to do this with more than one possible configuration of packages would be a bonus (for example, installing dipy / nipy wheels from some given location as part of the job initiation). 

I can imagine that others will want to do something similar with their own packages, and that it might be useful to extend the configuration in due course to something more general, like a .travis.yml setup.

Omar - is that right on the basic requirements?

**Comment by omarocegueda** [comment]
Hello everyone!
If we want to do a *full* evaluation of the module in terms of performance,
we would need to run the 306 registrations *for each metric, *and we would
need to use a large number of iterations. In that case it would be about
(50 minutes) * (306 volume pairs) * (3 metrics) for SyN. For affine
registration I need to see how slow is our implementation but I think it
should be about (5 minutes) * (306 volume pairs) * (1 metric). If we want
to do a quick run just to verify that our performance didn't change
recently, we can reduce the number of iterations and go to about 15 minutes
for SyN and perhaps to 2 minutes for affine. Right now, we need 2 GB of
memory for SyN, but I haven't checked for affine.

On Fri, Feb 27, 2015 at 3:26 PM, Matthew Brett <notifications@github.com>
wrote:

> Right - I was thinking that we would need some system which would run
> something like - say 50 jobs at a time? Omar - is your estimate then 15
> minutes * 306 for total CPU minutes? What CPUs are you running on? What
> kind of memory do they need?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/issues/441#issuecomment-76474692>.
>



-- 
"Cada quien es dueño de lo que calla y esclavo de lo que dice"
-Proverbio chino.
"We all are owners of what we keep silent and slaves of what we say"
-Chinese proverb.

http://www.cimat.mx/~omar

**Comment by matthew-brett** [comment]
So - JB - Chris - does the INCF have access to cluster time of this order that is it prepared to share?

Or should we see if we can negotiate something elsewhere?

**Comment by omarocegueda** [comment]
Yes, exactly, if we have storage for the data/results and some cpus to run
the experiments, we solve the original issue (being able to test that our
performance didn't decrease). I think the intention of Stefan was more like
Chris Filo described: having a service for people that want to evaluate
their own module, or ours with a custom configuration, but I agree that it
would be a lot of work (I would be fantastic, though!)

On Fri, Feb 27, 2015 at 4:12 PM, Matthew Brett <notifications@github.com>
wrote:

> Omar knows this much better than me, but I believe our immediate need is
> to be able to run scripts, say hosted on github, that will execute a
> multi-CPU job on the a given standard set of data.
>
> So yes, we would need:
>
> 1) somewhere to put the original data where the cluster can see it (and it
> would be better if this was also public;
> 2) CPUs on which to run the scripts, perhaps every week or so;
> 3) short term hosting of the output registrations and registration
> measures, which we could then download somewhere.
>
> Being able to do this with more than one possible configuration of
> packages would be a bonus (for example, installing dipy / nipy wheels from
> some given location as part of the job initiation).
>
> I can imagine that others will want to do something similar with their own
> packages, and that it might be useful to extend the configuration in due
> course to something more general, like a .travis.yml setup.
>
> Omar - is that right on the basic requirements?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/issues/441#issuecomment-76482419>.
>



-- 
"Cada quien es dueño de lo que calla y esclavo de lo que dice"
-Proverbio chino.
"We all are owners of what we keep silent and slaves of what we say"
-Chinese proverb.

http://www.cimat.mx/~omar

**Comment by omarocegueda** [comment]
sorry, I meant "It" would be fantastic!  =s

On Fri, Feb 27, 2015 at 8:57 PM, Jesus-Omar Ocegueda-Gonzalez <
jomaroceguedag@gmail.com> wrote:

> Yes, exactly, if we have storage for the data/results and some cpus to run
> the experiments, we solve the original issue (being able to test that our
> performance didn't decrease). I think the intention of Stefan was more like
> Chris Filo described: having a service for people that want to evaluate
> their own module, or ours with a custom configuration, but I agree that it
> would be a lot of work (I would be fantastic, though!)
>
> On Fri, Feb 27, 2015 at 4:12 PM, Matthew Brett <notifications@github.com>
> wrote:
>
>> Omar knows this much better than me, but I believe our immediate need is
>> to be able to run scripts, say hosted on github, that will execute a
>> multi-CPU job on the a given standard set of data.
>>
>> So yes, we would need:
>>
>> 1) somewhere to put the original data where the cluster can see it (and
>> it would be better if this was also public;
>> 2) CPUs on which to run the scripts, perhaps every week or so;
>> 3) short term hosting of the output registrations and registration
>> measures, which we could then download somewhere.
>>
>> Being able to do this with more than one possible configuration of
>> packages would be a bonus (for example, installing dipy / nipy wheels from
>> some given location as part of the job initiation).
>>
>> I can imagine that others will want to do something similar with their
>> own packages, and that it might be useful to extend the configuration in
>> due course to something more general, like a .travis.yml setup.
>>
>> Omar - is that right on the basic requirements?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/nipy/dipy/issues/441#issuecomment-76482419>.
>>
>
>
>
> --
> "Cada quien es dueño de lo que calla y esclavo de lo que dice"
> -Proverbio chino.
> "We all are owners of what we keep silent and slaves of what we say"
> -Chinese proverb.
>
> http://www.cimat.mx/~omar
>



-- 
"Cada quien es dueño de lo que calla y esclavo de lo que dice"
-Proverbio chino.
"We all are owners of what we keep silent and slaves of what we say"
-Chinese proverb.

http://www.cimat.mx/~omar

**Comment by chrisfilo** [comment]
I cannot speak on behalf of INCF, but I doubt they will be able to provide
such infrastructure. I'll check out other solutions though.
sorry, I meant "It" would be fantastic! =s

On Fri, Feb 27, 2015 at 8:57 PM, Jesus-Omar Ocegueda-Gonzalez <
jomaroceguedag@gmail.com> wrote:

> Yes, exactly, if we have storage for the data/results and some cpus to run
> the experiments, we solve the original issue (being able to test that our
> performance didn't decrease). I think the intention of Stefan was more
like
> Chris Filo described: having a service for people that want to evaluate
> their own module, or ours with a custom configuration, but I agree that it
> would be a lot of work (I would be fantastic, though!)
>
> On Fri, Feb 27, 2015 at 4:12 PM, Matthew Brett <notifications@github.com>
> wrote:
>
>> Omar knows this much better than me, but I believe our immediate need is
>> to be able to run scripts, say hosted on github, that will execute a
>> multi-CPU job on the a given standard set of data.
>>
>> So yes, we would need:
>>
>> 1) somewhere to put the original data where the cluster can see it (and
>> it would be better if this was also public;
>> 2) CPUs on which to run the scripts, perhaps every week or so;
>> 3) short term hosting of the output registrations and registration
>> measures, which we could then download somewhere.
>>
>> Being able to do this with more than one possible configuration of
>> packages would be a bonus (for example, installing dipy / nipy wheels
from
>> some given location as part of the job initiation).
>>
>> I can imagine that others will want to do something similar with their
>> own packages, and that it might be useful to extend the configuration in
>> due course to something more general, like a .travis.yml setup.
>>
>> Omar - is that right on the basic requirements?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/nipy/dipy/issues/441#issuecomment-76482419>.
>>
>
>
>
> --
> "Cada quien es dueño de lo que calla y esclavo de lo que dice"
> -Proverbio chino.
> "We all are owners of what we keep silent and slaves of what we say"
> -Chinese proverb.
>
> http://www.cimat.mx/~omar
>



-- 
"Cada quien es dueño de lo que calla y esclavo de lo que dice"
-Proverbio chino.
"We all are owners of what we keep silent and slaves of what we say"
-Chinese proverb.

http://www.cimat.mx/~omar

—
Reply to this email directly or view it on GitHub
<https://github.com/nipy/dipy/issues/441#issuecomment-76506935>.

**Comment by jbpoline** [comment]
Hi,

Let me check - I know they had access to a KI cluster, but since Romans
left I am not sure what is the situation now.

cheers
JB

**Comment by matthew-brett** [comment]
Just for reference, the full SyN run with 3 metrics is about 20 USD per run on Amazon AWS with a t2.small instance (1 CPU, 2GB of memory):

50 * 306 * 3 / 60. * 0.026  = 19.89

**Comment by jbpoline** [comment]
Hi,

with some delay, this is Linda (INCF) answer to it - so there may be a
possibility (to be investigated) - Let me know if you want to try that route

cheers
JB



On Tue, Mar 10, 2015 at 8:43 AM, Linda Lanyon <lindal@incf.org> wrote:
JB, sorry for the lack of response on this - I think it must have hit my
spam filter as I was unaware until Mathew altered me.

Anyway, INCF can offer supercomputer access - we have part-ownership of a
server hosted at KTH university and available to any of the INCF community:
https://www.incf.org/resources/access-to-super-computing-resources

If this is of interest, Mikael Djurfeldt (INCF-KTH) will be able to help
and Mathew can also help facilitate.

Best wishes


Linda

----------------------------------------------------------------------------------------------------
Linda Lanyon PhD CEng CITP MBCS
Executive Director
International Neuroinformatics Coordinating Facility
Karolinska Institutet
Nobels väg 15A, 171 77 Stockholm, Sweden
Phone: +46 8 524 86919 | Fax: +46 8 524 87094 | www.incf.org

**Comment by stefanv** [comment]
You can probably get away a bit cheaper if you use spot instances:

http://aws.amazon.com/ec2/purchasing-options/spot-instances/

**Comment by jbpoline** [comment]
Well, I would expect the incf access to be free until a certain limit -
will inquire more ...
cheers
JB

**Comment by stefanv** [comment]
I think we also have computing facilities on campus that is free for our
types of use cases.

**Comment by jbpoline** [comment]
that would probably be easier !

**Comment by matthew-brett** [comment]
Sure - but I think we should set this going on more than one cluster, to 

a) make sure we can do it easily;
b) be confident that this algorithm is running automatically for the next 5 years or so.

So, if the INCF cluster has free time, I think we should work on setting that up.   Of course we can also (in addition) set this up on the Berkeley neuro cluster.

**Comment by jbpoline** [comment]
Will pursue this then !

**Comment by matthew-brett** [comment]
JB - any progress here?

**Comment by jbpoline** [comment]
So, there seems indeed a serious cluster at KI and apparently we could use
some time of it - amount uncertain.  HBM will be an opportunity to talk to
Linda and others about it. If things need to be done before (I know it's
been some time already), I'll see what I can do.
cheers
JB

**Comment by matthew-brett** [comment]
Omar - sorry to return to this after such a long time.  Do you have these scripts anywhere on github, that I could look at?  

I'm currently planning teaching for the coming term, and I'd like to give people a go with a standard registration pipeline - was thinking to start with your scripts.

**Comment by omarocegueda** [comment]
Hi Matthew!,
sure!, my scripts are in this public repository:
https://github.com/omarocegueda/experiments/tree/master/experiments/registration

The first thing that needs to be done to run the experiments is to create a text file containing the full paths of the input images and ground truth segmentations. For example, for the LPBA40 dataset, I called my file "names_lpba40_full.txt", the first line contains the following file names separated with a blank space:
```
/home/omar/data/LPBA40/delineation_space/S01/S01_strip.img
/home/omar/data/LPBA40/delineation_space/S01/S01_strip_seg.img
```
the first name is the input image, the second is the ground truth segmentation.

The procedure I follow to run a full evaluation of a method is the following:

1. I create a sub directory inside the "experiments/registration" folder for my experiment. 

2. Run the split_[method].py from the newly created folder. For example, I like to put the names file in the same folder of my experiment. To run dipy's syn with the cc metric, do:
`python ../split_syn_cc.py s names.txt`. The 's' parameter, tells the script to "split" the dataset into individual jobs for the cluster, one job for each pair of images (it will run n*(n-1) registrations, not half of them because each image is run once as static image and once as moving image). If you haven't run affine registration for this dataset before, it will warn you that affine registration will be run first, and then it will run syn. If you have the affine results already, you can create a folder inside your experiment folder called `affine` and soft link the affine matrices (text files) from there. The naming pattern I use for the matrices is: `{moving}_{static}Affine.txt` where `moving` and `static` are the prefixes of the image names up to the first "dot" (I think in the code I just remove at most two dots from the right... sorry I wanted to fix that but I never did =( ).

3. Once the split finishes, you will find n*(n-1) folders. Each of this folders contains three subfolders:
`reference`, `target` and `warp`, (reference means static, target means moving... sorry, it's a legacy naming convention I was using before writing the python module). Each job folder also contains `dipyreg.py`, which runs the actual registration, `jaccard_pairs.lst` which tells `dipyreg.py` what pairs of images it must compute the jaccard indices from, and `job_syn_cc.sh` which is the actual cluster job, configuring the required cpu resources and calling `dipyreg.py` with the appropriate parameters.

4. To submit all jobs to the queue, run the split script again, this time with the `u` parameter ("sUbmit"):
`python ../split_syn_cc.py u` (no need to provide the `names.txt` folder)

5. After all jobs have successfully finished, you can call the split script again, this time with the `o` parameter ("cOllect"). This will create a `results` folder and will move all the required results there. Right now, I do not ask `dipyreg.py` to save all diffeomorphic maps because they consume **a lot** of space, so I just save the warped images. The Jaccard indices are written in text files called `jaccard_{warped}_{groud truth}.txt`, its just a list of indices, one per each segmentation label. If you want to compute the Jaccard indices for two or more different segmentations of the same image (e.g., one segmentation according to tissue type, another according to anatomical regions, etc.) you just need to include them in your `names.txt` file: in each line write the input image, and the segmentations separated by blank space, it will add those segmentations to the `jaccard_pairs.lst` and compute the evaluation for all of them. For example, for IBSR, each line of my `names_ibsr_full.txt` file contains four names:
```
/home/omar/data/IBSR_nifti_stripped/IBSR_01/IBSR_01_ana_strip.nii.gz
/home/omar/data/IBSR_nifti_stripped/IBSR_01/IBSR_01_segTRI_fill_ana.nii.gz
/home/omar/data/IBSR_nifti_stripped/IBSR_01/IBSR_01_segTRI_ana.nii.gz
/home/omar/data/IBSR_nifti_stripped/IBSR_01/IBSR_01_seg_ana.nii.gz
```
because IBSR provides two tissue-type segmentations (CSF, GM, WM, one original segmentation and one corrected by filling "holes" in the CSF that were marked as backround) and one anatomical region -based segmentation (31+ anatomical regions), and I like to compute all (cheap) evaluations after each (expensive) registration.

I just realized, it's a lot of details!, I hope it helps, please let me know if you need any help, Matthew!

**Comment by matthew-brett** [comment]
Thanks - that's very helpful.

Are the affine and SyN default parameters the same as for some standard ANTS paper?   I'm wondering whether I can justify these defaults to skeptical students...

**Comment by omarocegueda** [comment]
No problem!, I'm glad it was helpful! =)
I took de parameters for affine registration from a sample shell script I found in the ANTS repository. The parameters for SyN were kindly provided by Dr. Rohlfing in a private conversation:
```
ANTS 3 -t SyN[0.25] -r Gauss[3,0] -m CC[${ref_hdr},${flt_hdr},1,4] -o ${out_xfm}
```
After several years of studying this problem, I still feel like a beginner, and I think I would still feel the same if I studied it for another decade... or two. The more I learned about it the more skeptical I became about all aspects of the algorithms. I think the amount of parameters is so large that it is virtually impossible to find optimal parameters even for a particular dataset like IBSR (there are a few parameters that are exposed to the user, but there are many more that are "hardcoded" within the algorithm), so I wouldn't try to convince skeptical students about the "optimality" of the parameters. I wouldn't try to convince them either about the correctness of the registration result: at the end, we are just optimizing an arbitrary matching functional and assuming the optimal point corresponds to correct alignment, I think there's no way to convince a skeptical person about this being "correct", but just "reasonable". Instead of convincing them, I would just encourage them to go deeper to understand the details of what the algorithms are doing. I think our module is small, transparent and well documented enough so the user can understand the details.

**Comment by matthew-brett** [comment]
Yes, I am sure you are right to be skeptical.  At least, I am skeptical too.

I didn't mean to imply that the students would think the parameters were 'correct', but only that they were in some (specified) sense 'standard'.  Do you think these parameters are in any sense 'standard' - if so - do you know how Dr Rohlfing came by them?

@binarybottle - do you have in mind some standard set of ANTS parameters that the students could start with?  Have they changed since your NeuroImage paper?

**Comment by binarybottle** [comment]
I would ask Brian Avants or Nick Tustison for guidance.  It might help to take a look at this article as well:

"A reproducible evaluation of ANTs similarity metric performance in brainimage registration"
https://mfr.osf.io/render?url=https://osf.io/z59ad/?action=download%26mode=render

**Comment by matthew-brett** [comment]
Hi Omar - sorry to resurrect this thread, but I am having some difficulty following your registration scripts.  How would I compile a dipy optimizer that matches:

```
ANTS 3 -t SyN[0.25] -r Gauss[3,0] -m CC[${ref_hdr},${flt_hdr},1,4] -o ${out_xfm}
```

?

**Comment by skoudoro** [comment]
@TomasGuija, please have a look
